Add ES6 module support

keego opened the pull request:
This adds support for ES6 modules (fixing #197) which are commonly used in modern javascript development practices. This uses the `pkg.module` route [suggested by the guys over at Rollup](https://github.com/rollup/rollup/wiki/pkg.module) which *seems* to be the most standard way to add ES6 module support without making breaking changes.

The main changes are an amendment to package.json:
```diff
   "main": "./src/Ros3D.js",
+  "module": "./src/Ros3D.esm.js",
```

And duplicating Ros3D.js with an appended `export default ROS3D` at the bottom.

I recognize this isn't very DRY so I'm open to alternative methods of re-exporting Ros3D.js but I figure this at least gets the ball rolling.

**UPDATE**: Changes required for this PR ended up growing quite a bit, checkout [my comment below](https://github.com/RobotWebTools/ros3djs/pull/214#issuecomment-381802250) for details.

**Notes on importing**: 

Importing as a module
```js
import * as ROS3D from 'ros3d'
// Depending on your module bundler, this may either target:
// - ros3d.cjs.js - an ES5 compliant CommonJS module
// - ros3d.esm.js - an ES5 compliant ES module
``` 

Importing into an html document (should be the same):
```html
<!-- unminified - easier build for debugging -->
<script src="path/to/ros3d/build/ros3d.js"></script>
<!-- minified - smaller build for production -->
<script src="path/to/ros3d/build/ros3d.min.js"></script>
```

Some additional examples have been added to show different import semantics as well.

Closes #197

keego commented:
Ah this is going to be less trivial than I thought. Seems like your guys' Grunt pipeline will need some adjustments too...

keego commented:
Okay so after doing some digging, I found some interesting results.

The general structure this project uses of implicitly depending on globals may be conducive to the early web days of global namespace clobbering, but makes it kind of sketchy to support ES6 modules without significant code changes and/or breaking backwards compatibility. Plus I'm curious how it affects performance as I'm not sure if the ES6 modules will be able to tree shake much.

That said, it turns out that its fairly simple to extend the current build process to have an additional build target that is ES6 compliant. Simply concatenating some imports, `build/ros3d.js`, and an 'export default ROS3D' actually gets us pretty far in terms of turning global dependencies into modular ones.

So while this isn't the prettiest solution, it shouldn't have any effect on current users or the current development effort. Changes required to keep supporting the ES6 module should be minimal, likely requiring little to no changes to ES6 wrapper / grunt config.

keego commented:
**Please don't merge this yet btw**

I'm referencing these changes in my current project but I want to make sure to work out any and all kinks beforehand

keego commented:
Yay more progress!

### Choosing a module bundler
Looking at Webpack vs Rollup, I think Rollup may be a better fit for bundling this library due to its simplicity and output bundle size.

(See  [this excellent SO answer](https://stackoverflow.com/questions/43219030/what-is-flat-bundling-and-why-is-rollup-better-at-this-than-webpack) and [this medium post](https://medium.com/webpack/webpack-and-rollup-the-same-but-different-a41ad427058c) for more info)

__TLDR on the articles / my reasoning:__
While Webpack has a lot more features, they're generally more helpful for web apps than libraries (HMR, lazy-loading modules, etc.). However, code-splitting (aka tree-shaking) would be nice to have for ros3d consumers, but I believe they *should* get that anyway since we'd be exporting an ES6 module and ES6 modules natively support code-splitting.

### What needs to change
Rollup has an [awesome online REPL](https://rollupjs.org/repl) which makes it easy to test things out. You can select the desired module format for the output bundle as well. "es" is shows the ES6 bundle output (which would be imported by web apps via npm) and "iife" demonstrates a browser compatible bundle for including via a `script` tag. 

[This example](https://rollupjs.org/repl?version=0.56.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMG1haW4uanMlNUNuJTVDbiUyRiolMjBERUZBVUxUJTIwRVhQT1JUUyU1Q24lMjAlMjAlMjBEZWZhdWx0JTIwZXhwb3J0cyUyMGZyb20lMjB0aGUlMjAnZW50cnklMjBtb2R1bGUnJTIwYXJlJTVDbiUyMCUyMCUyMGV4cG9ydGVkJTIwZnJvbSUyMHRoZSUyMGJ1bmRsZSUyMColMkYlNUNuJTVDbmltcG9ydCUyMCcuJTJGUm9zM0QnJTVDbmltcG9ydCUyMCcuJTJGSGlnaGxpZ2h0ZXInJTVDbiUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJSb3MzRC5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlMkYlMkYlMjBSb3MzRC5qcyU1Q24lNUNudmFyJTIwUk9TM0QlMjAlM0QlMjBST1MzRCUyMCU3QyU3QyUyMCU3QiU1Q24lMjAlMjBSRVZJU0lPTiUyMCUzQSUyMCcwLjE4LjAnJTVDbiU3RCUzQiU1Q24lNUNuJTJGJTJGJTIwTWFya2VyJTIwdHlwZXMlNUNuUk9TM0QuTUFSS0VSX0FSUk9XJTIwJTNEJTIwMCUzQiU1Q25ST1MzRC5NQVJLRVJfQ1VCRSUyMCUzRCUyMDElM0IlNUNuJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMkhpZ2hsaWdodGVyLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMEhpZ2hsaWdodGVyLmpzJTVDbiU1Q25ST1MzRC5IaWdobGlnaHRlciUyMCUzRCUyMGZ1bmN0aW9uKG9wdGlvbnMpJTIwJTdCJTVDbiUyMCUyMG9wdGlvbnMlMjAlM0QlMjBvcHRpb25zJTIwJTdDJTdDJTIwJTdCJTdEJTNCJTVDbiUyMCUyMHRoaXMubW91c2VIYW5kbGVyJTIwJTNEJTIwb3B0aW9ucy5tb3VzZUhhbmRsZXIlM0IlNUNuJTIwJTIwdGhpcy5ob3Zlck9ianMlMjAlM0QlMjAlN0IlN0QlM0IlNUNuJTVDbiUyMCUyMCUyRiUyRiUyMGJpbmQlMjB0aGUlMjBtb3VzZSUyMGV2ZW50cyU1Q24lMjAlMjB0aGlzLm1vdXNlSGFuZGxlci5hZGRFdmVudExpc3RlbmVyKCdtb3VzZW92ZXInJTJDJTIwdGhpcy5vbk1vdXNlT3Zlci5iaW5kKHRoaXMpKSUzQiU1Q24lMjAlMjB0aGlzLm1vdXNlSGFuZGxlci5hZGRFdmVudExpc3RlbmVyKCdtb3VzZW91dCclMkMlMjB0aGlzLm9uTW91c2VPdXQuYmluZCh0aGlzKSklM0IlNUNuJTdEJTNCJTVDbiU1Q24lMkYqKiU1Q24lMjAqJTIwQWRkJTIwdGhlJTIwY3VycmVudCUyMHRhcmdldCUyMG9mJTIwdGhlJTIwbW91c2VvdmVyJTIwdG8lMjB0aGUlMjBob3ZlciUyMGxpc3QuJTVDbiUyMColNUNuJTIwKiUyMCU0MHBhcmFtJTIwZXZlbnQlMjAtJTIwdGhlJTIwZXZlbnQlMjB0aGF0JTIwY29udGFpbnMlMjB0aGUlMjB0YXJnZXQlMjBvZiUyMHRoZSUyMG1vdXNlb3ZlciU1Q24lMjAqJTJGJTVDblJPUzNELkhpZ2hsaWdodGVyLnByb3RvdHlwZS5vbk1vdXNlT3ZlciUyMCUzRCUyMGZ1bmN0aW9uKGV2ZW50KSUyMCU3QiU1Q24lMjAlMjB0aGlzLmhvdmVyT2JqcyU1QmV2ZW50LmN1cnJlbnRUYXJnZXQudXVpZCU1RCUyMCUzRCUyMGV2ZW50LmN1cnJlbnRUYXJnZXQlM0IlNUNuJTdEJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMlJPUzNEJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=) demonstrates the current module definition semantics and how they make bundling difficult (global depencies aside). 

[This example](https://rollupjs.org/repl?version=0.56.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMG1haW4uanMlNUNuJTVDbiUyRiolMjBERUZBVUxUJTIwRVhQT1JUUyU1Q24lMjAlMjAlMjBEZWZhdWx0JTIwZXhwb3J0cyUyMGZyb20lMjB0aGUlMjAnZW50cnklMjBtb2R1bGUnJTIwYXJlJTVDbiUyMCUyMCUyMGV4cG9ydGVkJTIwZnJvbSUyMHRoZSUyMGJ1bmRsZSUyMColMkYlNUNuJTVDbmV4cG9ydCUyMColMjBmcm9tJTIwJy4lMkZSb3MzRCclNUNuZXhwb3J0JTIwKiUyMGZyb20lMjAnLiUyRkhpZ2hsaWdodGVyJyU1Q24lMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyUm9zM0QuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyJTJGJTJGJTIwUm9zM0QuanMlNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwUkVWSVNJT04lMjAlM0QlMjAnMC4xOC4wJyUzQiU1Q24lNUNuJTJGJTJGJTIwTWFya2VyJTIwdHlwZXMlNUNuZXhwb3J0JTIwY29uc3QlMjBNQVJLRVJfQVJST1clMjAlM0QlMjAwJTNCJTVDbmV4cG9ydCUyMGNvbnN0JTIwTUFSS0VSX0NVQkUlMjAlM0QlMjAxJTNCJTVDbiUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJIaWdobGlnaHRlci5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlMkYlMkYlMjBIaWdobGlnaHRlci5qcyU1Q24lNUNuZXhwb3J0JTIwY2xhc3MlMjBIaWdobGlnaHRlciUyMCU3QiU1Q24lNUNuJTVDdGNvbnN0cnVjdG9yKG9wdGlvbnMpJTIwJTdCJTVDbiU1Q3QlNUN0b3B0aW9ucyUyMCUzRCUyMG9wdGlvbnMlMjAlN0MlN0MlMjAlN0IlN0QlM0IlNUNuJTVDdCU1Q3R0aGlzLm1vdXNlSGFuZGxlciUyMCUzRCUyMG9wdGlvbnMubW91c2VIYW5kbGVyJTNCJTVDbiU1Q3QlNUN0dGhpcy5ob3Zlck9ianMlMjAlM0QlMjAlN0IlN0QlM0IlNUNuJTVDbiU1Q3QlNUN0JTJGJTJGJTIwYmluZCUyMHRoZSUyMG1vdXNlJTIwZXZlbnRzJTVDbiU1Q3QlNUN0dGhpcy5tb3VzZUhhbmRsZXIuYWRkRXZlbnRMaXN0ZW5lcignbW91c2VvdmVyJyUyQyUyMHRoaXMub25Nb3VzZU92ZXIuYmluZCh0aGlzKSklM0IlNUNuJTVDdCU1Q3R0aGlzLm1vdXNlSGFuZGxlci5hZGRFdmVudExpc3RlbmVyKCdtb3VzZW91dCclMkMlMjB0aGlzLm9uTW91c2VPdXQuYmluZCh0aGlzKSklM0IlNUNuJTVDdCU3RCU1Q24lNUNuJTVDdCUyRioqJTVDbiU1Q3QlMjAqJTIwQWRkJTIwdGhlJTIwY3VycmVudCUyMHRhcmdldCUyMG9mJTIwdGhlJTIwbW91c2VvdmVyJTIwdG8lMjB0aGUlMjBob3ZlciUyMGxpc3QuJTVDbiU1Q3QlMjAqJTVDbiU1Q3QlMjAqJTIwJTQwcGFyYW0lMjBldmVudCUyMC0lMjB0aGUlMjBldmVudCUyMHRoYXQlMjBjb250YWlucyUyMHRoZSUyMHRhcmdldCUyMG9mJTIwdGhlJTIwbW91c2VvdmVyJTVDbiU1Q3QlMjAqJTJGJTVDbiU1Q3Rvbk1vdXNlT3ZlcihldmVudCklMjAlN0IlNUNuJTVDdCU1Q3R0aGlzLmhvdmVyT2JqcyU1QmV2ZW50LmN1cnJlbnRUYXJnZXQudXVpZCU1RCUyMCUzRCUyMGV2ZW50LmN1cnJlbnRUYXJnZXQlM0IlNUNuJTVDdCU3RCU1Q24lN0QlNUNuJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMlJPUzNEJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=) shows how modern JavaScript libraries are written and bundled. I believe this is how we should be doing things.

@sevenbitbyte @jihoonl Thoughts on moving forward with the above mentioned format?


Sidenote: Rollup doesn't natively transpile ES6 to ES5, a necessary step for supporting older browsers. However,  [rollup-plugin-babel](https://github.com/rollup/rollup-plugin-babel) solves this problem for us :)

AustinDeric commented:
@keego, this is great work thanks!  I tried to test this implementation, as i am trying to do something similar.  I am building a react app with webpack.  

My steps are to install this repo 
```
npm install https://github.com/keego/ros3djs.git#develop
```

Then, i import this ros3d node module into the react app (which is very similar to the examples, for the purpose of testing this code):
https://gist.github.com/AustinDeric/1a4c2b9bb6bae68eab230c33d276d7d5

the console log of the ROSLIB.Ros object looks great and connects to my ros bridge.  

However the ROS3D.Viewer object produces the following error:
'''
TypeError: _ros3d2.default.Viewer is not a constructor
'''

keego commented:
@AustinDeric My forked branch is still broken and will be for a while. I wanted to do some investigating before getting it ES6 compatible, especially in light of seeing just how much modification is necessary to make the change. I'll be posting here when I get it working.

How is ros3djs currently used with rclnodejs? Do you have some example code? I don't have much to test my changes against.

Also, I'd like to clarify I'll be moving forward with __Rollup__ _not Webpack_.

jihoonl commented:
Too sad that I don't have enough time for proper review at the moment.. @sevenbitbyte or @viktorku  could you guys take a look?

BTW, I would love to add ES6 module support. Thanks for the PR!

sevenbitbyte commented on Gruntfile.js:
```diff
@@ -6,6 +6,10 @@ module.exports = function(grunt) {
         dest : './build/ros3d.js'
+      },
+      build_esm: {
+        src  : [ './src-esm/ESMHeader.js', './build/ros3d.js', './src-esm/ESMFooter.js'],
```
Not required but might be more concise to keep all code in `./src` directory rather than creating a new one. Do you need a new directory to avoid a conflict with the existing grunt job for instance?

keego replied:
Nope, `src-esm` will be removed. I was toying with the idea of wrapping the original src instead of rewriting it. I imagine we should just move to writing ES6 only now?

sevenbitbyte commented on Gruntfile.js:
```diff
@@ -32,7 +37,7 @@ module.exports = function(grunt) {
         src: './build/ros3d.js',
         dest: './build/ros3d.min.js'
-      }
+      },
```
Watch out for dangling comma... or maybe linter is fine with this

keego replied:
Current jshint rules are unopinionated about dangling commas. After looking around I haven't seen dangling commas elsewhere so I'll stick to the defacto no-trailing-comma style for this codebase.

keego commented:
Sidenote: @jihoonl Where is `ROS3D.COLLADA_LOADER_2` defined? I've found [various references to it](https://github.com/RobotWebTools/ros3djs/blob/develop/src/interactivemarkers/InteractiveMarker.js#L27) but can't seem to find it defined anywhere.

jihoonl commented:
COLLADA_LOADER_2 is not available anymore. It is replaced by https://github.com/RobotWebTools/ros3djs/blob/develop/examples/js/ColladaLoader.js while three.js code base update in #202

keego commented:
@jihoonl May I go ahead and remove remaining references to `ROS3D.COLLADA_LOADER_2` (such as [in InteractiveMarker](https://github.com/RobotWebTools/ros3djs/blob/develop/src/interactivemarkers/InteractiveMarker.js#L27)) then?
i.e.
from:
```js
var loader = options.loader || ROS3D.COLLADA_LOADER_2;
```
to:
```js
var loader = options.loader;
```

jihoonl commented:
go ahead.

jihoonl commented:
@keego hello. Any updates on this?

sevenbitbyte commented:
@keego is on holiday this week so would expect a formal update around the 22nd. But @jihoonl looks like he's pretty close to solving this - https://github.com/keego/ros3djs/commits/ES6_ONLY

jihoonl commented:
Good. Thanks for the update

keego commented:
# Good news everyone!

![2009-02-27-professor_farnsworth](https://user-images.githubusercontent.com/5094178/38451108-9b4eaea6-39de-11e8-9b91-c5cb0ec686bf.jpg)

I was able to transpile ros3djs into an es6 module and get it properly importing into a single page application!

I had to make a few semantic changes to the original source code to get things to run properly:
- Removed references to COLLADA_LOADER_2
- Removed explicit call to a grandparent's constructor in InteractiveMarker
- Made Particles *explicitly* derive from THREE.Object3D
- Rewrote Particles' method signatures to be in a format consistent with the rest of the codebase
- Refactored updateMatrixWorld to be more statically analyzable
- Added a missing super constructor call
- Removed an assignment to a read-only property causing a runtime error
- Moved all super constructor calls to precede any use of `this`

### Shims

Also, in order to allow for the extension of `THREE`, I added a small shim that re-exports three in a way that allows it to be extended without mutating the `module` object (which would throw an error).
An example of how this works in practice can be found in `MeshResource.js` which now begins with:

```js
import THREE from '../../shims/three/core.js';
import '../../shims/three/STLLoader.js';
import '../../shims/three/ColladaLoader.js';
```

`STLLoader` and `ColladaLoader` are just local copies of those extensions which can get manually updated at will. While this isn't the most elegant solution, it does allow ros3djs to encapsulate any dependencies on three.js or three.js extensions.

### Dependencies

The new es6 source code bundles have *almost* all their dependencies rolled up into them. The exception is roslibjs, which is unable to be bundled at this time (they use runtime exports which aren't statically analyzable).

### Examples

I added a few examples (with READMEs included) that demonstrate how to use the different build targets of ros3djs.
- I added a [html-import](https://github.com/keego/ros3djs/tree/5343795b5d8640f1f75dbc369f0d3982041c7adb/examples/html-import) subfolder that shows how to use the ros3djs ECMAScript module in-browser.
- I added a [single-page-application](https://github.com/keego/ros3djs/tree/5343795b5d8640f1f75dbc369f0d3982041c7adb/examples/single-page-application) subfolder that shows how to use the ros3djs ECMAScript module in a single page application (or any general module bundler for that matter).

Also, I think it may be nice to add examples for Angular and React to the single-page-application section to help meet people where they're at.

### Moving forward

I have the code sitting in two branches of my ros3djs fork:

1. [ES5_WITH_TRANSPILER](https://github.com/keego/ros3djs/tree/ES5_WITH_TRANSPILER)
This branch contains the original source code (plus the "few semantic changes" listed above) as well as the transpiler that transpiles the source code to ES6. This serves as a point of compatibility with outstanding PRs.

2. [ES6_ONLY](https://github.com/keego/ros3djs/tree/ES6_ONLY)
This branch contains just the transpiled source code with the transpiler ripped out. This is what we'd be ideally targeting for future code changes.

@jihoonl @sevenbitbyte __What are your guys' thoughts on moving forward with the following?__:
1. Pull both branches into the ros3djs repo as 2 new branches
2. Target ES6_ONLY for new PRs
3. Finish merging any outstanding code changes into develop
4. Port develop changes into ES6_ONLY
5. Merge ES6_ONLY into develop

AustinDeric commented:
**@keego this is really great!  This PR will have a historical impact on opensource robotics, congrats!**  

I tested the ES6_ONLY branch at [commit 6dc97](https://github.com/keego/ros3djs/commit/6dc976d17348950fcd306e5d73295b4f4fdccbbd)

For my single page app using react js here is a gist for the react component test code:
<script src="https://gist.github.com/AustinDeric/f6b7ccaf24cffedea7d6c5a5eb61aa2b.js"></script>

I had add the following to your code in order for it to work because your code does use but does not import ROSLIB:
```
import ROSLIB from 'roslib';
```

There might be an additional error because the URDF is not correct:
![image](https://user-images.githubusercontent.com/5349043/39199196-7d6f99ee-47ae-11e8-851c-785d4edcbe37.png)

This is the URDF (from the same rosbridge as used above) when viewed from the current (global) ROS3D code:
![image](https://user-images.githubusercontent.com/5349043/39199088-48834802-47ae-11e8-9a03-d2eb8f168085.png)

keego commented:
@mbredif @AustinDeric Glad to help! 😃 

I was having some weird issues with importing ROSLIB so I'm glad simply adding the imports worked for you guys.

I just pushed commits to both branches that update the ROSLIB import semantics.

@AustinDeric I wasn't able to test all the examples so thanks for the feedback on the URDF demo. I'm using ROS Kinetic but I believe some of the examples require ROS Groovy packages. Hmm, if you remove this line from the urdf.html example:

keego commented:
@mbredif @AustinDeric Awesome, glad to hear! 😃 

I was having some weird issues with importing ROSLIB so I'm glad simply adding the imports worked for you guys.

I just pushed commits to both branches that update the ROSLIB import semantics.

@AustinDeric I wasn't able to test all the examples so thanks for the feedback on the URDF demo. I'm using ROS Kinetic but I believe some of the examples require ROS Groovy packages. Hmm, if you remove this line from the urdf.html example:
```html
<script src="js/ColladaLoader.js"></script> <!-- override the default ColladaLoader -->
```
does it appear to have the same issues as your react app?

It looks like the ColladaLoader extension used by `MeshResource` (and transitively `Urdf`) is the official  offical one. I just updated the ColladaLoader shim to use the ros3djs fork instead and pushed the changes to both branches. Mind giving your react gist another test?

keego commented:
Glad to help @jihoonl. I made some changes to the build process to get it working with the current CI pipeline (which also lead to GitHub auto closing this issue, oops). It looks like you guys are building with a very old version of Node, which isn't liking the transpiler.

**Any chance you could update the Travis CI config to build with the latest stable Node version (v8.9.4)?**

AustinDeric commented:
- I ended up not needing the import ROSLIB statement, i added webpack to keego's work and build the .cjs file and that works really nice.  The I will make a PR with this work once this PR get merged.
- Broken URDF, i am still looking into this.  I currently have no global includes in the html file (yay!).  It could very well be specific to my setup.  If not i will make a dockerfile and package to share with the team.
- Which file is intended to be the compiled output to ES5?  It looks like it would be ./build/ros3d.cjs.js but that file is not ES5 compliant and cannot be used as an external dependency (hence my webpack fix).

keego commented:
@AustinDeric `ros3d.cjs.js` wasn't es5 compliant. Turns out it still used some es6 features despite being in commonjs format. I pushed some changes that make sure the commonjs module is in es5.

AustinDeric commented:
@keego, i saw that.  I am working off your ES6_ONLY branch and transpiling with webpack and babel into es5.  Should i switch to the develop branch?

keego commented:
@jihoonl Any chance you could update the Travis CI config to build with the latest stable Node version (v8.9.4)? The (much) older version of Node being used doesn't like the transpiler.

keego commented:
@jihoonl just realized that the Travis CI config is literally in the repo :face_palm: so I just specified travis ci to use 8.9.4 instead of 0.10. All lights green 👍

jihoonl commented:
thanks for long waiting.. though I couldn't test it properly, i decided to merge it to move forward.

jihoonl commented:
Also,. I have created `develop-es6` branch as you requested.

keego commented:
Oh great, thanks! Yeah no worries, I figured you were busy.

keego commented:
@jihoonl I documented the transition process (#228) and started the PR for updating `develop-es6` (#229). Should be good to go at your own pace 👍 feel free to mention or email me for questions / comments / help.

Cheers :beers:
